### ISSUE TITLE
AttributeError: 'AsgiMiddleware' object has no attribute 'handle_async' when running locally

### ISSUE THREAD

**Issue by oli-hall** [Azure-Samples/fastapi-on-azure-functions#7]
Hi, I'm trying to run this locally (OSX BigSur v11.7) to get to grips with a basic example before integrating functions into my main FastApi project. I've set up the CLI integration as [suggested in Azure docs](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-cli-python?tabs=azure-cli%2Cbash&pivots=python-mode-configuration#configure-your-local-environment), and am running:
```
> func --version
4.0.4829
> az --version
azure-cli                         2.42.0

core                              2.42.0
telemetry                          1.0.8

Dependencies:
msal                              1.20.0
azure-mgmt-resource             21.1.0b1
```

The repo is cloned to a directory with a virtualenv running Python 3.7.5. When I run `func start`, the worker starts as expected, but when I send a `GET` request to `http://localhost:7071/sample`, I get an exception to the effect that `handle_async` isn't a recognised attribute of `AsgiMiddleware`:

```
[2022-11-01T13:04:51.693Z] Executed 'Functions.WrapperFunction' (Failed, Id=7f65d33d-f786-44e0-ad6a-a173327fe49a, Duration=157ms)
[2022-11-01T13:04:51.694Z] System.Private.CoreLib: Exception while executing function: Functions.WrapperFunction. System.Private.CoreLib: Result: Failure
[2022-11-01T13:04:51.694Z] Exception: AttributeError: 'AsgiMiddleware' object has no attribute 'handle_async'
[2022-11-01T13:04:51.694Z] Stack:   File "/usr/local/Cellar/azure-functions-core-tools@4/4.0.4829/workers/python/3.7/OSX/X64/azure_functions_worker/dispatcher.py", line 445, in _handle__invocation_request
[2022-11-01T13:04:51.694Z]     fi_context, fi.func, args
[2022-11-01T13:04:51.694Z]   File "/usr/local/Cellar/azure-functions-core-tools@4/4.0.4829/workers/python/3.7/OSX/X64/azure_functions_worker/dispatcher.py", line 698, in _run_async_func
[2022-11-01T13:04:51.694Z]     context, func, params
[2022-11-01T13:04:51.694Z]   File "/usr/local/Cellar/azure-functions-core-tools@4/4.0.4829/workers/python/3.7/OSX/X64/azure_functions_worker/extension.py", line 147, in get_async_invocation_wrapper
[2022-11-01T13:04:51.694Z]     result = await function(**args)
[2022-11-01T13:04:51.694Z]   File "/<user-dir>/fastapi-on-azure-functions/WrapperFunction/__init__.py", line 21, in main
[2022-11-01T13:04:51.694Z]     return await func.AsgiMiddleware(app).handle_async(req, context)
```

I reverted to the previous `handle` code (using `nest_asyncio`), and that works fine. I suspect some library mismatch, as PyCharm can find the `handle_async` method just fine, but any ideas? I'm not sure on the inner workings of this stuff, so really unsure how to dig deeper in debugging this.

**Comment by oli-hall** [comment]
A bit more digging, and it seems that although `pip` has installed `1.12.0`, the version being executed is running `1.11.3b3`, which lacks the `handle_async` method.

```
→ pip freeze | grep azure
azure-functions==1.12.0
→ python
Python 3.7.5 (default, Aug 19 2020, 17:32:28) 
[Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import azure.functions as func
>>> func.__version__
'1.12.0'
```

I added the following into the `main()` function in `WrapperFunction.__init__.py`:
```
async def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
    print("version:", func.__version__)
    print(dir(func.AsgiMiddleware))
    return func.AsgiMiddleware(app).handle(req, context)
```

When run and a request is made, it dumps out the following:
```
[2022-11-01T13:26:45.430Z] version: 1.11.3b3
[2022-11-01T13:26:45.430Z] ['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_handle', '_logger', '_usage_reported', 'handle']
```

**Comment by slyons** [comment]
@oli-hall Mind sharing your workaround? I'm having the same problem

**Comment by nformbiz** [comment]
I'm seeing the exact same error both locally and when deployed to azure functions.  In azure functions the version is 1.12 but still see the error "AsgiMiddleware' object has no attribute 'handle_async"

**Comment by oli-hall** [comment]
@slyons Sure, I basically just used the behaviour from [the previous commit](https://github.com/Azure-Samples/fastapi-on-azure-functions/blob/c11a418450182b8aa2a492977fd22e9ad239ab18/WrapperFunction/__init__.py), before the shift to `handle_async`. Basically, that involves the following changes:

1. install `nest_asyncio` (`pip install nest_asyncio`, add it to `requirements.txt`, etc).
2. In your FastAPI module (your equivalent to `WrapperFunction/__init__.py` here), add the following:
```
import nest_asyncio

nest_asyncio.apply()
```
3. Change the invocation of `handle_async` to `handle`:
```
return func.AsgiMiddleware(app).handle(req, context)
```

It's a deprecated method, so you'll get a warning, but I've been running it fine both locally and deployed without issue.

**Comment by alijewers** [comment]
+1 to this, seeing the exact same issue

**Comment by CribberSix** [comment]
+1 again.  I used oli-hall's solution for now...

**Comment by ma-ikea** [comment]
It's working if you are using python 3.10

**Comment by karolpawlowski** [comment]
> It's working if you are using python 3.10

Anyone from maintainers could confirm what @ma-ikea wrote? Do I have to migrate to python 3.10 to use `handle_async`?

**Comment by tonybaloney** [comment]
Set the environment variable 

PYTHON_ISOLATE_WORKER_DEPENDENCIES=1

This will make sure your code installs the latest version so this function will not be bundled with the host dependencies.

**Comment by tonybaloney** [comment]
No, you don't need to migrate to Python 3.10. This method works on 3.7+. 
I'll investigate some of the comments in this thread.

**Comment by pamelafox** [comment]
Just going to give a few more tips as I've been playing with this.

## Local dev

If you're running into this issue when running Azure Functions Core Tools **locally**, here are two approaches.

1. Specify the env variable when running the command:

  ```
  PYTHON_ISOLATE_WORKER_DEPENDENCIES=1 func host start
  ```

2. Update tasks.json to specify the env variable, and use the VS Code task:

  https://github.com/pamelafox/fastapi-on-azure-functions/blob/azd-support/.vscode/tasks.json#L12

## Deployment

If you're running into it once it's **deployed**, here are two approaches.

1. Update the env variables in the Configuration in the Portal:

  ![Screenshot 2022-11-23 at 1 06 56 PM](https://user-images.githubusercontent.com/297042/203646224-afbcad99-302e-4510-94b3-8551030a23f2.png)

2. Specify the env variable in a Bicep file and deploy with the Azure Dev CLI:

  https://github.com/pamelafox/fastapi-on-azure-functions/blob/azd-support/infra/resources.bicep#L105

  Here's a fork of this repo can be deployed with "azd up":
  https://github.com/pamelafox/fastapi-on-azure-functions

**Comment by ma-ikea** [comment]
> > It's working if you are using python 3.10
> 
> Anyone from maintainers (@tonybaloney?) could confirm what @ma-ikea wrote? Do I have to migrate to python 3.10 to use `handle_async`?

I tested on my local environment with python versions 3.9 and 3.10. With 3.9 I get same error as others, but when I switch to 3.10 it's gone :)

**Comment by karolpawlowski** [comment]
@tonybaloney @pamelafox Thanks for the hint with `PYTHON_ISOLATE_WORKER_DEPENDENCIES=1`

Works even with Python 3.8.

**Comment by nformbiz** [comment]
I can confirm on Windows 10 that this is all working locally with Azure Functions Core Tools version 4.0.4915.

In the Azure Functions server adding PYTHON_ISOLATE_WORKER_DEPENDENCIES=1 to settings solved this issue immediately with no code changes.

**Comment by JamesPrenticez** [comment]
Thanks @pamelafox =) 

`"PYTHON_ISOLATE_WORKER_DEPENDENCIES": 1`

I found adding this to the local.setting.json worked instead of tasks.json.

**Comment by oli-hall** [comment]
This thread/issue is fairly settled now, but just thought I'd add this note here in case anyone else has the same behaviour as I encountered just now. Today all my deployed Azure Functions apps stopped working, and all calls to them just hung until they reached the maximum timeout (in my case, 5 minutes).

I was still running my original work-around, but on a hunch updated to the above fix (`PYTHON_ISOLATE_WORKER_DEPENDENCIES=1` with `handle_async()`) and everything resolved itself and all calls are now working just fine. Thanks again to those who recommended this as a fix!
